File: src/research_artifact_uploader/manifest.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List

import yaml

DEFAULT_ARCHIVE_OUTPUT_DIR = ".rau/archives"
DEFAULT_ARCHIVE_FORMAT = "tar.gz"
DEFAULT_ARCHIVE_COMPRESSION_LEVEL = 3
DEFAULT_RECORDS_JSONL = "docs/upload_records.jsonl"
DEFAULT_RECORDS_MARKDOWN = "docs/upload_records.md"

DEFAULT_OSS_BUCKET = "luyukuan-research"
DEFAULT_OSS_REGION = "cn-shanghai"
DEFAULT_OSS_ENDPOINT = "oss-accelerate.aliyuncs.com"


@dataclass
class RunConfig:
    name: str
    project: str
    tags: List[str]


@dataclass
class ArtifactConfig:
    name: str
    path: str
    type: str
    required: bool


@dataclass
class ArchiveConfig:
    output_dir: str
    format: str
    compression_level: int
    include_manifest: bool
    include_git_info: bool
    include_metadata: bool


@dataclass
class OssConfig:
    bucket: str
    region: str
    endpoint: str
    remote_dir: str | None


@dataclass
class RecordsConfig:
    jsonl: str
    markdown: str


@dataclass
class Manifest:
    run: RunConfig
    artifacts: List[ArtifactConfig]
    archive: ArchiveConfig
    oss: OssConfig
    records: RecordsConfig

    def to_dict(self) -> Dict[str, Any]:
        return {
            "run": {
                "name": self.run.name,
                "project": self.run.project,
                "tags": list(self.run.tags),
            },
            "artifacts": [
                {
                    "name": artifact.name,
                    "path": artifact.path,
                    "type": artifact.type,
                    "required": artifact.required,
                }
                for artifact in self.artifacts
            ],
            "archive": {
                "output_dir": self.archive.output_dir,
                "format": self.archive.format,
                "compression_level": self.archive.compression_level,
                "include_manifest": self.archive.include_manifest,
                "include_git_info": self.archive.include_git_info,
                "include_metadata": self.archive.include_metadata,
            },
            "oss": {
                "bucket": self.oss.bucket,
                "region": self.oss.region,
                "endpoint": self.oss.endpoint,
                "remote_dir": self.oss.remote_dir,
            },
            "records": {
                "jsonl": self.records.jsonl,
                "markdown": self.records.markdown,
            },
        }


def _expect_mapping(payload: Dict[str, Any], key: str):
    value = payload.get(key)
    if not isinstance(value, dict):
        raise ValueError(f"{key} must be a mapping")
    return value


def _expect_list(payload: Dict[str, Any], key: str):
    value = payload.get(key)
    if value is None:
        return []
    if not isinstance(value, list):
        raise ValueError(f"{key} must be a list")
    return value


def _expect_str(value: Any, key: str) -> str:
    if not isinstance(value, str):
        raise ValueError(f"{key} must be a string")
    return value


def _expect_bool(value: Any, key: str, default: bool) -> bool:
    if value is None:
        return default
    if isinstance(value, bool):
        return value
    raise ValueError(f"{key} must be true or false")


def _artifact_type(value: Any) -> str:
    if value not in {"file", "directory", "glob"}:
        raise ValueError("artifact type must be file, directory, or glob")
    return value


def parse_manifest(path: Path) -> Manifest:
    raw = path.read_text(encoding="utf-8")
    payload = yaml.safe_load(raw)
    if not isinstance(payload, dict):
        raise ValueError("manifest must be YAML mapping")

    run = _expect_mapping(payload, "run")
    name = _expect_str(run.get("name"), "run.name")
    project = _expect_str(run.get("project"), "run.project")
    tags_raw = run.get("tags", [])
    if not isinstance(tags_raw, list) or any(not isinstance(x, str) for x in tags_raw):
        raise ValueError("run.tags must be list of strings")

    run_cfg = RunConfig(name=name, project=project, tags=list(tags_raw))

    artifacts_raw = _expect_list(payload, "artifacts")
    artifacts: List[ArtifactConfig] = []
    for i, item in enumerate(artifacts_raw):
        if not isinstance(item, dict):
            raise ValueError(f"artifacts[{i}] must be mapping")
        artifacts.append(
            ArtifactConfig(
                name=_expect_str(item.get("name"), f"artifacts[{i}].name"),
                path=_expect_str(item.get("path"), f"artifacts[{i}].path"),
                type=_artifact_type(item.get("type")),
                required=_expect_bool(item.get("required"), f"artifacts[{i}].required", False),
            )
        )

    archive_cfg_raw = _expect_mapping(payload, "archive")
    archive_cfg = ArchiveConfig(
        output_dir=_expect_str(archive_cfg_raw.get("output_dir", DEFAULT_ARCHIVE_OUTPUT_DIR), "archive.output_dir"),
        format=_expect_str(archive_cfg_raw.get("format", DEFAULT_ARCHIVE_FORMAT), "archive.format"),
        compression_level=int(archive_cfg_raw.get("compression_level", DEFAULT_ARCHIVE_COMPRESSION_LEVEL)),
        include_manifest=_expect_bool(archive_cfg_raw.get("include_manifest"), "archive.include_manifest", True),
        include_git_info=_expect_bool(archive_cfg_raw.get("include_git_info"), "archive.include_git_info", True),
        include_metadata=_expect_bool(archive_cfg_raw.get("include_metadata"), "archive.include_metadata", True),
    )
    if archive_cfg.format != "tar.gz":
        raise ValueError("only tar.gz archive format is supported")
    if not 1 <= archive_cfg.compression_level <= 9:
        raise ValueError("archive.compression_level must be in [1,9]")

    oss_cfg_raw = _expect_mapping(payload, "oss")
    oss_cfg = OssConfig(
        bucket=_expect_str(oss_cfg_raw.get("bucket", DEFAULT_OSS_BUCKET), "oss.bucket"),
        region=_expect_str(oss_cfg_raw.get("region", DEFAULT_OSS_REGION), "oss.region"),
        endpoint=_expect_str(oss_cfg_raw.get("endpoint", DEFAULT_OSS_ENDPOINT), "oss.endpoint"),
        remote_dir=oss_cfg_raw.get("remote_dir"),
    )
    if not isinstance(oss_cfg.remote_dir, str) or not oss_cfg.remote_dir:
        raise ValueError("oss.remote_dir is required")

    records_cfg_raw = _expect_mapping(payload, "records")
    records_cfg = RecordsConfig(
        jsonl=_expect_str(records_cfg_raw.get("jsonl", DEFAULT_RECORDS_JSONL), "records.jsonl"),
        markdown=_expect_str(records_cfg_raw.get("markdown", DEFAULT_RECORDS_MARKDOWN), "records.markdown"),
    )

    return Manifest(
        run=run_cfg,
        artifacts=artifacts,
        archive=archive_cfg,
        oss=oss_cfg,
        records=records_cfg,
    )

File: src/manifest.rs
use std::{fs, path::Path};

use serde::{Deserialize, Serialize};
use serde_yaml::Value;

pub const DEFAULT_ARCHIVE_OUTPUT_DIR: &str = ".rau/archives";
pub const DEFAULT_ARCHIVE_FORMAT: &str = "tar.gz";
pub const DEFAULT_ARCHIVE_COMPRESSION_LEVEL: u32 = 3;
pub const DEFAULT_RECORDS_JSONL: &str = "docs/upload_records.jsonl";
pub const DEFAULT_RECORDS_MARKDOWN: &str = "docs/upload_records.md";
pub const DEFAULT_OSS_BUCKET: &str = "luyukuan-research";
pub const DEFAULT_OSS_REGION: &str = "cn-shanghai";
pub const DEFAULT_OSS_ENDPOINT: &str = "oss-accelerate.aliyuncs.com";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Manifest {
    pub run: RunConfig,
    #[serde(default)]
    pub artifacts: Vec<ArtifactConfig>,
    #[serde(default)]
    pub archive: ArchiveConfig,
    #[serde(default)]
    pub oss: OssConfig,
    #[serde(default)]
    pub records: RecordsConfig,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunConfig {
    pub name: String,
    pub project: String,
    #[serde(default)]
    pub tags: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ArtifactConfig {
    pub name: String,
    pub path: String,
    #[serde(rename = "type")]
    pub artifact_type: String,
    #[serde(default)]
    pub required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ArchiveConfig {
    #[serde(default = "default_archive_output_dir")]
    pub output_dir: String,
    #[serde(default = "default_archive_format")]
    pub format: String,
    #[serde(default = "default_archive_compression_level")]
    pub compression_level: u32,
    #[serde(default = "default_true")]
    pub include_manifest: bool,
    #[serde(default = "default_true")]
    pub include_git_info: bool,
    #[serde(default = "default_true")]
    pub include_metadata: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OssConfig {
    #[serde(default = "default_bucket")]
    pub bucket: String,
    #[serde(default = "default_region")]
    pub region: String,
    #[serde(default = "default_endpoint")]
    pub endpoint: String,
    pub remote_dir: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecordsConfig {
    #[serde(default = "default_records_jsonl")]
    pub jsonl: String,
    #[serde(default = "default_records_markdown")]
    pub markdown: String,
}

fn default_true() -> bool {
    true
}

fn default_archive_output_dir() -> String {
    DEFAULT_ARCHIVE_OUTPUT_DIR.to_string()
}
fn default_archive_format() -> String {
    DEFAULT_ARCHIVE_FORMAT.to_string()
}
fn default_archive_compression_level() -> u32 {
    DEFAULT_ARCHIVE_COMPRESSION_LEVEL
}
fn default_records_jsonl() -> String {
    DEFAULT_RECORDS_JSONL.to_string()
}
fn default_records_markdown() -> String {
    DEFAULT_RECORDS_MARKDOWN.to_string()
}
fn default_bucket() -> String {
    DEFAULT_OSS_BUCKET.to_string()
}
fn default_region() -> String {
    DEFAULT_OSS_REGION.to_string()
}
fn default_endpoint() -> String {
    DEFAULT_OSS_ENDPOINT.to_string()
}

impl Default for ArchiveConfig {
    fn default() -> Self {
        Self {
            output_dir: default_archive_output_dir(),
            format: default_archive_format(),
            compression_level: default_archive_compression_level(),
            include_manifest: true,
            include_git_info: true,
            include_metadata: true,
        }
    }
}

impl Default for RecordsConfig {
    fn default() -> Self {
        Self {
            jsonl: default_records_jsonl(),
            markdown: default_records_markdown(),
        }
    }
}

impl Default for OssConfig {
    fn default() -> Self {
        Self {
            bucket: default_bucket(),
            region: default_region(),
            endpoint: default_endpoint(),
            remote_dir: String::new(),
        }
    }
}

pub fn parse_manifest(path: &Path) -> Result<Manifest, String> {
    let raw = fs::read_to_string(path).map_err(|err| format!("cannot read manifest: {err}"))?;
    let payload: Value = serde_yaml::from_str(&raw).map_err(|err| format!("invalid manifest yaml: {err}"))?;

    let mapping = payload
        .as_mapping()
        .ok_or_else(|| "manifest root must be a mapping".to_string())?;

    let mut m: Manifest = serde_yaml::from_value(Value::Mapping(mapping.clone()))
        .map_err(|err| format!("invalid manifest schema: {err}"))?;

    if m.run.name.trim().is_empty() {
        return Err("run.name is required".to_string());
    }
    if m.run.project.trim().is_empty() {
        return Err("run.project is required".to_string());
    }

    if m.archive.format != "tar.gz" {
        return Err("only tar.gz archive format is supported".to_string());
    }
    if !(1..=9).contains(&m.archive.compression_level) {
        return Err("archive.compression_level must be in [1,9]".to_string());
    }

    if m.oss.remote_dir.trim().is_empty() {
        return Err("oss.remote_dir is required".to_string());
    }

    let allowed_types = ["file", "directory", "glob"];
    for (i, artifact) in m.artifacts.iter().enumerate() {
        if !allowed_types.contains(&artifact.artifact_type.as_str()) {
            return Err(format!(
                "artifacts[{i}].type must be file, directory, or glob"
            ));
        }
        if artifact.name.trim().is_empty() {
            return Err(format!("artifacts[{i}].name must not be empty"));
        }
        if artifact.path.trim().is_empty() {
            return Err(format!("artifacts[{i}].path must not be empty"));
        }
    }

    if m.records.jsonl.trim().is_empty() {
        m.records.jsonl = DEFAULT_RECORDS_JSONL.to_string();
    }
    if m.records.markdown.trim().is_empty() {
        m.records.markdown = DEFAULT_RECORDS_MARKDOWN.to_string();
    }

    Ok(m)
}

impl Manifest {
    pub fn to_map(&self) -> Result<String, String> {
        serde_yaml::to_string(self).map_err(|err| format!("serialize manifest: {err}"))
    }
}
